tests/components/matrix/test_commands.py

```diff
@@ -0,0 +1,155 @@
+        return {
+            "sender": "@SomeUser:example.com",
+            "room": self.room_id,
+        } | self.expected_event_data_extra
+
+
+room_message_base = partial(
+    RoomMessageText,
+    formatted_body=None,
+    format=None,
+    source={
+        "event_id": "fake_event_id",
+        "sender": "@SomeUser:example.com",
+        "origin_server_ts": 123456789,
+    },
+)
+word_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="!WordTrigger arg1 arg2"),
+    expected_event_data_extra={
+        "command": "WordTriggerEventName",
+        "args": ["arg1", "arg2"],
+    },
+)
+expr_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="My name is FakeName"),
+    expected_event_data_extra={
+        "command": "ExpressionTriggerEventName",
+        "args": {"name": "FakeName"},
+    },
+)
+word_command_subset = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="!WordTriggerSubset arg1 arg2"),
+    expected_event_data_extra={
+        "command": "WordTriggerSubsetEventName",
+        "args": ["arg1", "arg2"],
+    },
+)
+expr_command_subset = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="Your name is FakeName"),
+    expected_event_data_extra={
+        "command": "ExpressionTriggerSubsetEventName",
+        "args": {"name": "FakeName"},
+    },
+)
+# Messages without commands should trigger nothing
+fake_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(body="This is not a real command!"),
+    expected_event_data_extra=None,
+)
+# Valid commands sent by the bot user should trigger nothing
+self_command_global = partial(
+    CommandTestParameters,
+    room_message=room_message_base(
+        body="!WordTrigger arg1 arg2",
+        source={
+            "event_id": "fake_event_id",
+            "sender": TEST_MXID,
+            "origin_server_ts": 123456789,
+        },
+    ),
+    expected_event_data_extra=None,
+)
+
+
+@pytest.mark.parametrize("room_id", [TEST_ROOM_A_ID, TEST_ROOM_B_ID, TEST_ROOM_C_ID])
+@pytest.mark.parametrize(
+    "partial_param",
+    [
+        word_command_global,
+        expr_command_global,
+        word_command_subset,
+        expr_command_subset,
+        fake_command_global,
+        self_command_global,
+    ],
+)
+async def test_commands(
+    hass: HomeAssistant,
+    matrix_bot: MatrixBot,
+    command_events,
+    partial_param,
+    room_id: RoomID,
+):
+    """Test that the configured commands are used correctly."""
+    command_param: CommandTestParameters = partial_param(room_id=room_id)
+    room = MatrixRoom(room_id=command_param.room_id, own_user_id=matrix_bot._mx_id)
+
+    await hass.async_start()
+    assert len(command_events) == 0
+    await matrix_bot._handle_room_message(room, command_param.room_message)
+    await hass.async_block_till_done()
+
+    match command_events:
```

Comment:
We should avoid if or match statements in a test.

Reply:
I'll remove the match statements, but I'm not sure how I can avoid at least a single `if`: based on the combination of the parameters for the message command, and the room_id, I have to dynamically check that there's either a single `Event` with the correct data, or that there's no such event at all.

I'll keep thinking about it as I make the change, but hopefully that'll work